src/jupyterlabcontroller/services/builder/lab.py

```diff
@@ -0,0 +1,821 @@
+        """Determine the URL of a newly-spawned lab.
+
+        The hostname and port are fixed to match the Kubernetes ``Service`` we
+        create, but the local part is normally determined by an environment
+        variable passed from JupyterHub.
+
+        Parameters
+        ----------
+        username
+            Username of lab user.
+        env
+            Environment variables from JupyterHub.
+
+        Returns
+        -------
+        str
+            URL of the newly-spawned lab.
+        """
+        namespace = self.namespace_for_user(username)
+        path = env["JUPYTERHUB_SERVICE_PREFIX"]
+        return f"http://lab.{namespace}:8888" + path
+
+    def build_lab(
+        self,
+        *,
+        user: GafaelfawrUserInfo,
+        lab: LabSpecification,
+        image: RSPImage,
+        secrets: dict[str, str],
+        pull_secret: V1Secret | None = None,
+    ) -> LabObjects:
+        """Construct the objects that make up a user's lab.
+
+        Parameters
+        ----------
+        username
+            Username of the user.
+        lab
+            Specification of the lab requested.
+        image
+            Image to use for the lab.
+        secrets
+            Dictionary of secrets to expose to the lab.
+        pull_secret
+            Optional pull secret for the lab pod.
+
+        Returns
+        -------
+        LabObjects
+            Kubernetes objects that make up the user's lab.
+        """
+        return LabObjects(
+            namespace=self._build_namespace(user.username),
+            config_maps=self._build_config_maps(user, lab, image),
+            network_policy=self._build_network_policy(user.username),
+            pvcs=self._build_pvcs(user.username),
+            quota=self._build_quota(user),
+            secrets=self._build_secrets(user.username, secrets, pull_secret),
+            service=self._build_service(user.username),
+            pod=self._build_pod(user, lab, image),
+        )
+
+    def build_lab_config_volumes(
+        self,
+        username: str,
+        volumes: list[LabVolume],
+        prefix: str = "",
+    ) -> list[LabVolumeContainer]:
+        """Construct volumes and mounts for volumes in the lab configuration.
+
+        Parameters
+        ----------
+        username
+            Name of user this lab is for, used to name the PVC objects.
+        volumes
+            Configured volumes. This is a parameter rather than being taken
+            straight from the lab configuration because this method is also
+            used when constructing init containers, which use separate volume
+            lists.
+        prefix
+            Prefix to prepend to all mount paths, if given.
+
+        Returns
+        -------
+        list of LabVolumeContainer
+            List of volumes and mounts.
+        """
+        results = []
+        pvc_count = 1
+        for storage in volumes:
+            is_readonly = storage.mode == FileMode.RO
```

Comment:
Obviously we don't need them, but a set of parens here would help readability for me.

Reply:
I tried, but black just removes them again.